HTTP Remote Adapter (closes #3760)

szareiangm opened the pull request:
This ticket aims to add a feature to adapters to make them configurable for the reasons discussed in  #3760.

We have added an example for the curious users of this adapter, too. We didn't know where to put it, so we guessed a candidate location, but feel free to tell us your opinion. 

CC: @nafid @dmatth

alexanderdean commented:
This is somewhat similar (although this is an adapter not enrichment) to: https://github.com/snowplow/snowplow/issues/2054

szareiangm commented:
@BenFradet Sorry about late response. I want to move this discussion forward. 

I think we can work on a vanilla HTTP version, probably with POST. Is that okay with you?

Are you sure about the gRPC? It comes with dependency on Proto3. https://grpc.io/docs/tutorials/basic/java.html Dependency on Protubuf is a whole new story, we have already Json/TSV serialization in the pipeline and you guys have tickets for Avro.

BenFradet commented:
Agreed, I think HTTP is fine and the most non-controversial for the moment

szareiangm commented:
we have changed it as we agreed. can you please have a look?

BenFradet commented:
cool thanks I'll have a look

BenFradet commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,103 @@
+      case Right(bodyAsString) =>
+        RemoteAdapter
+          .deserializeFromBase64(bodyAsString)
+          .asInstanceOf[Either[List[String], List[RawEvent]]] match {
```
I'm not too sure about the input/output of the adapter, could we find a format better-suited for http?

szareiangm replied:
What is your suggestion? The choice is based on the fact that input is an unstructured text and output should be in format of `RawEvent` to be compatible with the next steps.

BenFradet replied:
In my mind the adapter could take care of transforming the input before forwarding the collector payload and then turn the result into a RawEvent when receiving it

szareiangm replied:
Got it. I think I should pull out some of the remote-adapter-example functionalities and put them here. So basically the steps would be these:
1) `RemoteAdapter` extracts the collector-specific information in the payload, and just sends the business related values/params to the webservice.
2) `Webservice` does the work to transfer the data to well-formed Snowplow event.
3) `RemoteAdapter` recives the payload and merges it with step1 data into a `RawEvent` payload.

BenFradet replied:
looks good on posting the collector payload, however why receive base64 and not directly json?

szareiangm replied:
encoding and binary data. We can try json and see what would happen. so shall I remove base64?

BenFradet replied:
yup I'd rather we go json out json in, created the repo: https://github.com/snowplow-incubator/remote-adapter-example/

szareiangm commented:
About moving the example to the incubator, how do we do it? @BenFradet

BenFradet commented:
I can create a repo for you :+1:

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/AdapterRegistry.scala:
```diff
@@ -57,6 +57,9 @@ object AdapterRegistry {
   }
 
+  //TODO make the following less of a hack
+  private val remoteAdapters = RemoteAdapters.createFromConfigFile(System.getProperty("remoteAdapterConfig"))
```
I will address it after the other comment

szareiangm commented:
@BenFradet The code is ready. I need to do one last end-to-end test for it and then we are good to go.

szareiangm commented:
Tests are successful. I think this is mergeable, too.

benjben commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,104 @@
+      else
+        Right(response.unsafeBody)
+
+    } catch {
```
Using `Try` and pattern matching would be nicer than having a big `try {} catch {}` and `if () else if () else ()`.
```scala
Try(sttp.readTimeout(timeout).body(body).post(uri"$requestUrl"))
  .flatMap(_.send()) match {
    case Success(response) if response.isClientError => ...
    case Success(response) if response.isServerError => ...
    case Success(response) => ...
    case Failure(err) => ...
  }
```

szareiangm commented on 3-enrich/beam-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/Enrich.scala:
```diff
@@ -135,6 +135,8 @@ object Enrich {
     .map { rawEvent =>
       cachedFiles()
       implicit val r = ResolverSingleton.get(resolver)
+      implicit val remoteAdapters = Map() // TODO: add later
```
#4021

szareiangm commented on 3-enrich/spark-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.spark/EnrichJob.scala:
```diff
@@ -118,6 +118,7 @@ object EnrichJob extends SparkJob {
     import singleton._
     val registry = RegistrySingleton.get(config.igluConfig, config.enrichments, config.local)
     val loader   = LoaderSingleton.get(config.inFormat).asInstanceOf[Loader[Any]]
+    implicit val remoteAdapters = Map() // TODO: add later
```
#4020

szareiangm commented:
@benjben Thank you for taking the time and reviewing the code. I will address the comments shortly. Most probably in the next 24 hours. 

About your design discussion:
I think we can look at this from another perspective, too: `RemoteAdapter` can be extended by more advanced RemoteAdapters and using inheritance we have the same list of extended classes objects. Probably by that time that we have those other RemoteAdapters, they would be called `AkkaRemoteAdapter` , `GrpcRemoteAdapter` and his would become `HttpReoteAdapter`. All of them share the same trait or base class. 

Basically, all of these remote adapters will have the same bare-bone structure of functionality: take the raw request, send it over, receive the formatted payload with with compatible data structure. The technology or network protocol should not deviate them from having same function interfaces.

How does it look like?

szareiangm commented on 3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/Enrich.scala:
```diff
@@ -245,4 +249,19 @@ a  * @param creds optionally necessary credentials to download the resolver
+    * @param remoteAdaptersConfig List of configuration per remote adapter
+    * @return Mapping of vender-version and the adapter assigned for it
+    */
+  def prepareRemoteAdapters(remoteAdaptersConfig: Option[List[RemoteAdapterConfig]]) = {
```
the way that the Spark configuration for the main application is passed and processed is somewhat different.

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,104 @@
+              })
+            }
+          } catch {
+            case  e: Exception => s"Unexpected response from remote $e".failNel
```
addressed

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,104 @@
+   *         Success, or a NEL of Failure Strings
+   */
+  def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents =
+    if (payload.body.isEmpty || payload.body.get == "") {
```
addressed

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,104 @@
+      val response = request.send()
+
+      if (response.isClientError)
+        Left(s"remote webservice rejected this request with a ${response.code} error! ('${response.body}')")
```
hmm.. the text is mis-leading. `isClientError` happens when the client has done something wrong

szareiangm commented on 3-enrich/scala-common-enrich/build.sbt:
```diff
@@ -19,7 +19,7 @@ lazy val root = project
   .settings(
     name        := "snowplow-common-enrich",
-    version     := "0.36.0",
+    version     := "0.36.0-SNAPSHOT",
```
This was just for the sake of the tests. removed.

szareiangm commented on 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/adapters/registry/RemoteAdapterSpec.scala:
```diff
@@ -0,0 +1,226 @@
+            write(Response(null, s"expecting a payload json but got a ${anythingElse.getClass}"))
+        }
+      } catch {
+        case e: Exception => write(Response(null, s"aack, sampleAdapter exception $e"))
```
addressed

benjben commented:
Hi @szareiangm !

_RemoteAdapter can be extended by more advanced RemoteAdapters_ : we can't know yet if other adapters would only need a URI and a timeout.

But even without talking about other kinds of adapters, I think that it's nicer to have one `EnrichmentRegistry` and one `AdapterRegistry` in the params. Also, I think that this is always good to avoid implicit parameters whenever it's possible. If you prefer to keep the current design, please put the map with the regular params rather than with implicit ones. You can also give it a default value (empty map) so that you don't need to specify it in case you don't have remote adapters.

szareiangm commented:
Sure. I think I can do it before noon EST today.

szareiangm commented:
I removed the implicit vals and changed the the design to have adapters. Some unit-tests do not run on my laptop due to some differences in the environment. I have to force-push a few time to get it done.

szareiangm commented:
Changes are done as you have requested @benjben 
@BenFradet can I ask if we could schedule this for R114, please?

BenFradet commented:
yup that's the plan :+1:

benjben commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,96 @@
+              }
+            } catch {
+              case NonFatal(e) => s"Unexpected response from remote $e".failNel
+            }
```
This `try catch` wraps many lines and different things (error and success), it would be nice to have more fine-grained control over the processing of the response.

szareiangm replied:
I have added a few things below. but I can imagine that would be it.

benjben commented on 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/adapters/registry/RemoteAdapterSpec.scala:
```diff
@@ -0,0 +1,231 @@
+
+  override def is = sequential ^ s2"""
+   This is a specification to test the RemoteAdapter functionality.
+   RemoteAdapter must return any events parsed by this local test enricher                 ${testWrapperLocal(e1)}
```
What do you mean by `this local test enricher`?

szareiangm replied:
It was adapter. mistyped.

benjben commented on 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/adapters/registry/RemoteAdapterSpec.scala:
```diff
@@ -0,0 +1,231 @@
+   RemoteAdapter must also return any other errors issued by this local enricher           ${testWrapperLocal(e3)}
+   RemoteAdapter must return any events parsed by an external test enricher                ${testWrapperExternal(e1ext)}
+   that external test enricher must treat an empty list as an error                        ${testWrapperExternal(e2ext)}
+   RemoteAdapter must also return any other errors issued by that external enricher        ${testWrapperExternal(e3ext)}
```
Few things:
- this -> `T`his and that -> `T`hat
- When you say enricher, or you actually talking about adapter?
- When you say errors, are you talking about the ones in the HTTP response?
- Empty list where?

I think that a few tests should be added:
- HTTP response contains string that is not a correct JSON
- HTTP response contains empty string
- HTTP response contains well-formatted JSON but without `events` and `error`
- HTTP response contains well-formatted JSON with `events` that is not a `List[Map[String, String]]`
- HTTP response contains well-formatted JSON with `events` that contains an empty list
- HTTP response contains status code other than 200

szareiangm replied:
I will finish the unit tests until tomorrow.

szareiangm replied:
done

benjben commented on 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/enrichments/registry/pii/PiiPseudonymizerEnrichmentSpec.scala:
```diff
@@ -135,8 +137,9 @@ class PiiPseudonymizerEnrichmentSpec extends Specification with ValidationMatche
     val input: ValidatedMaybeCollectorPayload = Some(collectorPayload).successNel
     val rrc                                   = new RepositoryRefConfig("test-schema", 1, List("com.snowplowanalytics.snowplow"))
     val repos                                 = TestResourcesRepositoryRef(rrc, "src/test/resources/iglu-schemas")
+    val adapterRegistry                       = new AdapterRegistry(Map.empty[(String, String), RemoteAdapter])
```
No need to specify `Map.empty[(String, String), RemoteAdapter]`.

szareiangm replied:
done

szareiangm commented:
@benjben  Thanks for your comments. I think I would be able to address them around the weekend.

szareiangm commented on 3-enrich/beam-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/singleton.scala:
```diff
@@ -61,6 +62,24 @@ object singleton {
+      if (instance == null) {
+        synchronized {
+          if (instance == null) {
+            instance = new AdapterRegistry(Map.empty)
```
done

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/AdapterRegistry.scala:
```diff
@@ -34,29 +33,8 @@ import registry._
-    val Vero            = "com.getvero"
-    val HubSpot         = "com.hubspot"
-  }
+class AdapterRegistry(remoteAdapters: Map[(String, String), RemoteAdapter]) {
```
done

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/AdapterRegistry.scala:
```diff
@@ -94,7 +72,34 @@ object AdapterRegistry {
       case (Vendor.HubSpot, "v1")               => HubSpotAdapter.toRawEvents(payload)
       case _ =>
-        s"Payload with vendor ${payload.api.vendor} and version ${payload.api.version} not supported by this version of Scala Common Enrich".failNel
+        remoteAdapters get ((payload.api.vendor, payload.api.version)) match {
```
done

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,96 @@
+    Try(sttp.readTimeout(timeout).body(body).post(uri"$requestUrl").send()) match {
+      case Success(response) =>
+        response.code match {
+          case 200 => Right(response.unsafeBody)
```
I simplified it

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,96 @@
+
+  val bodyMissingErrorText = "missing payload body"
+
+  private lazy val log = LoggerFactory.getLogger(getClass)
```
done

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,96 @@
+              case NonFatal(e) => s"Unexpected response from remote $e".failNel
+            }
+
+          case ng => s"Unexpected response from remote $remoteUrl $ng".failNel
```
removed

szareiangm commented:
@benjben The comments are addressed.

BenFradet commented on 3-enrich/scala-common-enrich/project/Dependencies.scala:
```diff
@@ -103,6 +106,7 @@ object Dependencies {
     val scalaWeather     = "com.snowplowanalytics"      %% "scala-weather"                 % V.scalaWeather
     val scalaj           = "org.scalaj"                 %% "scalaj-http"                   % V.scalaj
     val gatlingJsonpath  = "io.gatling"                 %% "jsonpath"                      % V.gatlingJsonpath
+    val sttp             = "com.softwaremill.sttp"      %% "core"                          % V.sttp
```
we already have scalaj on the classpath, I'd be in favour of using it instead of introducing a new dependency.

szareiangm replied:
removed it, but had to add timeout to HttpClient class

BenFradet commented on 3-enrich/scala-common-enrich/project/Dependencies.scala:
```diff
@@ -87,6 +89,7 @@ object Dependencies {
     val mysqlConnector   = "mysql"                      %  "mysql-connector-java"          % V.mysqlConnector
     val jaywayJsonpath   = "com.jayway.jsonpath"        %  "json-path"                     % V.jaywayJsonpath
     val iabClient        = "com.snowplowanalytics"      %  "iab-spiders-and-robots-client" % V.iabClient
+    val typesafeConfig   = "com.typesafe"               %  "config"                        % V.typesafeConfig
```
why is typesafe config required?

szareiangm replied:
removed. it was for the last iteration

BenFradet commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,112 @@
+package com.snowplowanalytics
```
missing apache header

BenFradet commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:
```diff
@@ -17,6 +17,7 @@ package common
 package enrichments
 
 // Joda
+import com.snowplowanalytics.snowplow.enrich.common.adapters.AdapterRegistry
```
why is this import needed?

szareiangm replied:
it was a mistake. sorry

BenFradet commented on 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/adapters/registry/RemoteAdapterSpec.scala:
```diff
@@ -0,0 +1,249 @@
+package com.snowplowanalytics.snowplow.enrich.common
```
missing apache header

BenFradet commented on 3-enrich/spark-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.spark/singleton.scala:
```diff
@@ -107,6 +108,25 @@ object singleton {
   }
 
+  /** Singleton for AdapterRegistry. */
+  object AdapterRegistrySingleton {
```
if it's not support for spark enrich, no need to define this singleton

BenFradet commented on 3-enrich/stream-enrich/examples/config.hocon.sample:
```diff
@@ -175,6 +175,16 @@ enrich {
   }
 
+  # The setting below requires an adapter being ready, i.e.: https://github.com/snowplow-incubator/remote-adapter-example
+  # remoteAdapters:[
```
:art: `remoteAdapters = [`

BenFradet commented on 3-enrich/stream-enrich/kinesis/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/KinesisEnrich.scala:
```diff
@@ -29,7 +29,6 @@ import scala.collection.JavaConverters._
 import scala.io.Source
 import scala.sys.process._
 import scala.util.Try
-
```
:art: why were those lines removed? we tend to separate java from scala from other imports

BenFradet commented on 3-enrich/beam-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/singleton.scala:
```diff
@@ -61,6 +62,23 @@ object singleton {
+  }
 
+  /** Singleton for AdapterRegistry. */
+  object AdapterRegistrySingleton {
```
if it's not supported by beam, no need to define this singleton

benjben commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala:
```diff
@@ -0,0 +1,96 @@
+    }
+
+  private def httpPost(requestUrl: String, body: String): Either[String, String] =
+    Try(sttp.readTimeout(timeout).body(body).post(uri"$requestUrl").send()) match {
```
Could you update this please?

szareiangm replied:
Good morning! sure. I need a few hours to address all of the comments. I will start with this one

szareiangm commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentRegistry.scala:
```diff
@@ -200,7 +200,7 @@ object EnrichmentRegistry {
 
   /**
-   * A list of all files required by enrichments in the registry.
+   * A list of all files required by enrichments in the enrichmentRegistry.
```
this was a mistake from my IDE, too. reverting.

szareiangm commented:
I am done addressing the comments. `beam` and `spark` modules fail because CI/CD compiles them with the same version of `common` and I have reverted my changes in them.

benjben commented:
Hi @szareiangm , thanks for addressing the comments. After updating `scala-common-enrich`, the enrich jobs need to be updated so that they still compile and pass the unit tests with the new version of the library. In your case it seems that you updated `stream-enrich` correctly, passing `adaptersRegistry` to `processEvents`, but this is not the case for `beam-enrich` and `spark-enrich`.

szareiangm commented:
Thank you! Sure. I am pushing the changes. Might take a few commits

szareiangm commented:
I addressed your comments but `beam` fails.
I compiled `master` branch and got the same error. It is not from my branch. The error message doesn't look very dev friendly to understand the reason..

benjben commented:
Thanks a lot for the changes @szareiangm !
Sorry for `beam-enrich` we could have told you, the issue had already been identified and fixed (see [here](https://github.com/snowplow/snowplow/commit/b0a14e5a2dff1a7070545201b1596fbe50debbed)) but not merged to `master` yet (it will be in next release).

I would like to add an integration test before the end of the week and then it's all good :+1:

szareiangm commented:
😄 I spent 3-4 hours digging reflections in `scio` with the debugger with different combinations of code changes. It was not fun honestly but that's okay, we are friends 👍 
I wish we could have this PR in the next release somehow, It has been a long journey!

benjben commented:
Oh no too bad, sorry about that!
Yes it will be in the next release!

benjben commented:
Superseded by #4076
